src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs

```diff
@@ -32,58 +33,120 @@ internal abstract partial class AbstractInitializeParameterCodeRefactoringProvid
-                method.ContainingType.TypeKind == TypeKind.Interface)
-            {
-                return;
-            }
+            parameterNodes = GetParameters(functionDeclaration, generator);
 
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-            if (CanOfferRefactoring(functionDeclaration, semanticModel, syntaxFacts, cancellationToken, out var blockStatementOpt))
+            foreach (var parameterNode in parameterNodes)
             {
-                // Ok.  Looks like a reasonable parameter to analyze.  Defer to subclass to 
-                // actually determine if there are any viable refactorings here.
-                context.RegisterRefactorings(await GetRefactoringsAsync(
-                    document, parameter, functionDeclaration, method, blockStatementOpt, cancellationToken).ConfigureAwait(false));
+                ++counter;
+
+                // Only offered when there isn't a selection, or the selection exactly selects a parameter name.
+                var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+                if (!context.Span.IsEmpty)
+                {
+                    var parameterName = syntaxFacts.GetNameOfParameter(parameterNode);
+                    if (parameterName == null || parameterName.Value.Span != context.Span)
+                    {
+                        continue;
+                    }
+                }
+
+                var parameterDefault = syntaxFacts.GetDefaultOfParameter(parameterNode);
+
+                // Don't offer inside the "=initializer" of a parameter
+                if (parameterDefault?.Span.Contains(position) == true)
+                {
+                    continue;
+                }
+
+                // we can't just call GetDeclaredSymbol on functionDeclaration because it could an anonymous function,
+                // so first we have to get the parameter symbol and then its containing method symbol
+                var parameter = (IParameterSymbol)semanticModel.GetDeclaredSymbol(parameterNode, cancellationToken);
+                if (parameter == null || parameter.Name == "")
+                {
+                    return;
+                }
+
+                var methodSymbol = (IMethodSymbol)parameter.ContainingSymbol;
+                if (methodSymbol.IsAbstract ||
+                    methodSymbol.IsExtern ||
+                    methodSymbol.PartialImplementationPart != null ||
+                    methodSymbol.ContainingType.TypeKind == TypeKind.Interface)
+                {
+                    continue;
+                }
+
+                if (CanOfferRefactoring(functionDeclaration, semanticModel, syntaxFacts, cancellationToken, out var blockStatementOpt))
+                {
+                    // Ok.  Looks like there is at least one reasonable parameter to analyze.  Defer to subclass to 
+                    // actually determine if there are any viable refactorings here.
```

Comment:
hmm... I still don't like the way this loop is constructed. Let chat offline to see if we can agree on something different :)